### PR TITLE
Print stats and separate setup and run from prove

### PR DESCRIPTION
**Pull request by Vimalakirti**
as the title stated

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -116,7 +152,16 @@ fn main() {
...
+  setup(&srs, &graph, &models);
+  let outputs = run(&inputs, &graph, &models);
+  prove(&srs, &inputs, outputs, &mut graph);
   verify(&srs, &graph);
+  // measure_proof_size
+  let _modelsEncSize = measure_file_size("modelsEnc");
+  let inputsEncSize = measure_file_size("inputsEnc");
+  let outputsEncSize = measure_file_size("outputsEnc");
+  let proofsSize = measure_file_size("proofs");
+  let totalSize = inputsEncSize + outputsEncSize + proofsSize;
```
Comment (Vimalakirti) [inline src/main.rs]: I think the proof size only includes the size of inputsEnc, outputsEnc, and proofs. But I may be wrong.

Comment (ddkang) [reply src/main.rs]: @liliatangxy can you check this

File: src/main.rs [hunk 2]
```diff
@@ -71,7 +101,11 @@ fn prove(srs: &SRS, inputs: &Vec<&ArrayD<Fr>>, graph: &mut Graph, models: &Vec<&
   let mut rng = StdRng::from_seed(buf);
 
   // Prove:
+  let start = std::time::Instant::now();
   let proofs = graph.prove(srs, &setups, &models, &inputs, &outputs, &mut rng);
+  let prove_time = start.elapsed();
+  stat_println!("Time taken to prove: {:?}", prove_time);
+  stat_println!("Total time taken to prove: {:?}", encode_time + prove_time);
```
Comment (Vimalakirti) [inline src/main.rs]: I think the proving time should include the time to msm I/O and to run prove. But I may be wrong.

File: src/main.rs [hunk 3]
```diff
@@ -21,16 +21,36 @@ mod ptau;
 mod tests;
 mod util;
 
-fn prove(srs: &SRS, inputs: &Vec<&ArrayD<Fr>>, graph: &mut Graph, models: &Vec<&ArrayD<Fr>>) {
-  // Run:
-  let outputs = graph.run(inputs, models);
+macro_rules! stat_println {
```
Comment (ddkang) [inline src/main.rs]: Aren't there good libraries for this? I believe plonky2/3 have good examples

**Review by ddkang — COMMENTED** [review]
I think we should use a library since the library can determine if it should print based on env variables, etc

File: src/main.rs (hunk 2 above)
Comment (ddkang) [reply src/main.rs]: What do you mean by MSM i/o? If we assume steady state, we don't need to time the time to load the SRS

Comment (Vimalakirti) [reply src/main.rs]: It does not include the time to load the SRS. It means we commit the input tensor, intermediate tensors, and the output tensor by msm them.
(I/O is ambiguous, I should say msm input and output tensors)

File: src/main.rs (hunk 3 above)
Comment (Vimalakirti) [reply src/main.rs]: I just checked Plonky2, they add a feature tag above their print function (i.e., `#[cfg(feature = "timing")]`). Do you think this simple fix could address this issue?

Comment (ddkang) [reply src/main.rs]: Don't they also have a way to preprint the timing?

File: src/main.rs (hunk 2 above)
Comment (ddkang) [reply src/main.rs]: Ah okay, yes we should include that but not for the one-time setup.

File: src/main.rs (hunk 3 above)
Comment (Vimalakirti) [reply src/main.rs]: What do you mean by "preprint"? Do you mean print all the timing results together?

Comment (ddkang) [reply src/main.rs]: typo - pretty print

Comment (Vimalakirti) [reply src/main.rs]: I checked Plonky2 and 3 again. I don't think they have pretty print. What they have above ours is the feature tag above print, and the use of debug print somewhere.

Comment (Vimalakirti) [reply src/main.rs]: But maybe we can use [this](https://crates.io/crates/prettytable-rs) to print them in an organizing way.

File: src/main.rs (hunk 1 above)
Comment (liliatangxy) [reply src/main.rs]: I think it should just contain proofs

File: src/main.rs (hunk 3 above)
Comment (liliatangxy) [reply src/main.rs]: Plonky2 has `TimingTree` to print out proving time breakdowns: https://github.com/0xPolygonZero/plonky2/blob/main/plonky2/src/util/timing.rs#L8

Comment (Vimalakirti) [reply src/main.rs]: Thanks! Let me check how to use this struct.

**Comment by Vimalakirti** [comment]
Thanks for all the valuable comments. The PR is now ready for reviews.
The new changes:
- adopted `plonky2` `TimingTree`, which eases us to show the proving time and its breakdown
- fixed proof size to be only for `proofs`
- use nightly build for GitHub actions because `plonky2` needs it.

**Review by liliatangxy — APPROVED** [review]

**Review by ddkang — APPROVED** [review]

**Pull request merged**